src/core/qgsproject.h

```diff
@@ -575,6 +576,54 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
      */
     QMap<QString, QgsMapLayer *> mapLayers() const;
 
+    /**
+     * Unzip a project
+     * \param filename The zip file to unzip
+     * \returns true if unzip is well performed, false otherwise
+     * \since QGIS 3.0
+     */
+    bool unzip( const QString &filename );
+
+    /**
+     * Unzip a project with the current zip filename
+     * \returns true if unzip is well performed, false otherwise
+     * \since QGIS 3.0
+     */
+    bool unzip();
+
+    /**
+     * Zip the project
+     * \param filename The zip filename
+     * \returns true if zip is well performed, false otherwise
+     * \since QGIS 3.0
+     */
+    bool zip( const QString &filename );
+
+    /**
+     * Zip the project with the current zip filename
+     * \returns true if zip is well performed, false otherwise
+     * \since QGIS 3.0
+     */
+    bool zip();
+
+    /**
+     * Returns the current zip filename or an empty string if none.
+     * \since QGIS 3.0
+     */
+    QString zipFileName() const;
```

Comment:
@nyalldawson 

> I'd suggest leaving QgsProject.fileName() and .fileInfo() as always returning paths for the original user-location of the project (regardless of whether it's a qgs or qgz file), and adding some other accessor which always returns the path to the original or extracted .qgs file.

I agree with you, I'm gonna do that.

@wonder-sk 

> in theory right after load it could be removed again...?

It is quite true... I'm going to take a look at it.